### PR TITLE
Migrate Continuous Documentation from Vercel to Readthedocs

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

PyGMT has been using Vercel (previously Zeit Now) since 2019, and that's been a great tool in checking out documentation previews on individual Pull Requests (PRs). However, the legacy service plan we're on will be deprecated on 12 May 2022, see https://vercel.com/support/articles/legacy-plans-deprecation

This Pull Request migrates the configuration to Read*the*Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions.

I've set up the new URL to point to https://pygmt-devs.readthedocs.io. @GenericMappingTools/pygmt-maintainers, for those of you who would like access to the readthedocs backend, please register for an account on https://readthedocs.org/accounts/signup, and ping me so that I can add you to the maintainer list.

Cc @meghanrjones @seisman, I might need help setting up the readthedocs webhooks integration on the pygmt repo (ref https://docs.readthedocs.io/en/stable/integrations.html)

References:
- https://blog.readthedocs.com/pull-request-builder-general-availability/
- https://docs.readthedocs.io/en/stable/config-file/v2.html#version
- https://docs.readthedocs.io/en/stable/guides/conda.html

Supersedes #344

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
> Cc @meghanrjones @seisman, I might need help setting up the readthedocs webhooks integration on the pygmt repo (ref [docs.readthedocs.io/en/stable/integrations.html](https://docs.readthedocs.io/en/stable/integrations.html))

Approved.

**Comment by seisman** [comment]
> This Pull Request migrates the configuration to Read_the_Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions. I've set up the new URL to point to [pygmt-devs.readthedocs.io](https://pygmt-devs.readthedocs.io/).

My concern is, we will have two domains that host the exact same contents, which may affect the  SEO of of main site.

**Comment by weiji14** [comment]
> > This Pull Request migrates the configuration to Read_the_Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions. I've set up the new URL to point to [pygmt-devs.readthedocs.io](https://pygmt-devs.readthedocs.io/).
> 
> My concern is, we will have two domains that host the exact same contents, which may affect the SEO of of main site.

I've set the canonical URL to point to https://pygmt.org following https://docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls. I think that should ensure that search engines point to the existing .org site?

**Comment by seisman** [comment]
> > > This Pull Request migrates the configuration to Read_the_Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions. I've set up the new URL to point to [pygmt-devs.readthedocs.io](https://pygmt-devs.readthedocs.io/).
> > 
> > 
> > My concern is, we will have two domains that host the exact same contents, which may affect the SEO of of main site.
> 
> I've set the canonical URL to point to [pygmt.org](https://pygmt.org) following [docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls](https://docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls). I think that should ensure that search engines point to the existing .org site?

Looks great.

**Comment by weiji14** [comment]
> > > > This Pull Request migrates the configuration to Read_the_Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions. I've set up the new URL to point to [pygmt-devs.readthedocs.io](https://pygmt-devs.readthedocs.io/).
> > > 
> > > 
> > > My concern is, we will have two domains that host the exact same contents, which may affect the SEO of of main site.
> > 
> > 
> > I've set the canonical URL to point to [pygmt.org](https://pygmt.org) following [docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls](https://docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls). I think that should ensure that search engines point to the existing .org site?
> 
> Looks great.

Hmm, we might need to modify the pygmt.org DNS record somehow... Now the readthedocs preview documentation link points to http://pygmt.org/en/vercel_to_readthedocs :sweat_smile:

**Comment by weiji14** [comment]
I'm going to be flat out busy for the next month (travel, workshops, etc), so I probably won't have time to finish this up. @GenericMappingTools/pygmt-maintainers can someone maybe take over this Vercel to Readthedocs migration PR since Vercel is expiring soon? If you sign up for a readthedocs account at https://readthedocs.org/accounts/login/, I can add you to the pygmt-dev maintainer list at https://readthedocs.org/dashboard/pygmt-dev/users

image

**Comment by seisman** [comment]
Please add me as a maintainer so that I can give it a try.

**Comment by weiji14** [comment]
> Please add me as a maintainer so that I can give it a try.

Ok, done.

**Comment by seisman** [comment]
> > > > > This Pull Request migrates the configuration to Read_the_Docs which is another web hosting platform. Note that this will only affect the redendered documentation previews for PRs. The documentation built on the main branch is still built and deployed using GitHub Actions. I've set up the new URL to point to [pygmt-devs.readthedocs.io](https://pygmt-devs.readthedocs.io/).
> > > > 
> > > > 
> > > > My concern is, we will have two domains that host the exact same contents, which may affect the SEO of of main site.
> > > 
> > > 
> > > I've set the canonical URL to point to [pygmt.org](https://pygmt.org) following [docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls](https://docs.readthedocs.io/en/stable/custom_domains.html#canonical-urls). I think that should ensure that search engines point to the existing .org site?
> > 
> > 
> > Looks great.
> 
> Hmm, we might need to modify the pygmt.org DNS record somehow... Now the readthedocs preview documentation link points to http://pygmt.org/en/vercel_to_readthedocs 😅

I think you were NOT reading the right documentation when setting the canonical URLs. The correct documentation is here: https://docs.readthedocs.io/en/stable/canonical-urls.html. I've removed the settings of custom domains in the Admin panel.

**Comment by seisman** [comment]
The preview documentation for this PR is available at: https://pygmt-dev--1859.org.readthedocs.build/en/1859/

If you look at the source of the website, you will see this line:
```
    <link rel="canonical" href="https://pygmt.org/latest/index.html" />
```
which sets the canonical URLs to `https://pygmt.org/latest/index.html`. This is done by setting `html_baseurl` to `https://pygmt.org/latest/` in `doc/conf.py`.

The documentation building is slower (~15 minutes) than Vercel (~10 minutes). 

There is still one problem on the API documentation pages (https://pygmt-dev--1859.org.readthedocs.build/en/1859/api/index.html). All the links are NOT clickable.

File: .readthedocs.yaml
Comment (weiji14) [inline .readthedocs.yaml:26]: Do you think we should just install the sphinx-related libraries needed for building the docs? We used to install everything on Vercel because it's quite fast, but Readthedocs is a bit slower it seems, so less libraries might speed things up.

Comment (seisman) [reply .readthedocs.yaml]: Done in d21acf9519c2b34fd2ad569014a86f0cb39a4c9a.

Slightly faster. Now it takes 12 minutes.

File: .github/workflows/docs_environment.yml
Comment (meghanrjones) [reply .github/workflows/docs_environment.yml:1]: inspiration from xarray - https://github.com/pydata/xarray/tree/main/ci/requirements

Comment (seisman) [reply .github/workflows/docs_environment.yml:1]: Sounds a good place to put all CI related files. 

Ping @weiji14 for comments on this.

Comment (weiji14) [reply .github/workflows/docs_environment.yml:1]: Ideally we would put all these 'docs' dependencies as an 'extras' in pyproject.toml/setup.py following https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies so that we could just `pip install .[docs]`. However, GMT would still need to come from conda-forge, unless #1853 becomes a thing :slightly_smiling_face:

Comment (weiji14) [reply .github/workflows/docs_environment.yml:1]: Actually, is the GMT binary needed to build the docs? Could we just mock it out? I recall @meghanrjones was able to use https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_mock_imports at https://github.com/pangeo-data/xbatcher/pull/62#issuecomment-1117833077, though not sure if it works in this case.

Comment (seisman) [reply .github/workflows/docs_environment.yml:1]: > Actually, is the GMT C library needed to build the docs? Could we just mock it out?

I don't think it's possible. We need to call the GMT C library to generate images (e.g., all the images generated by sphinx-gallery).

File: ci/requirements/docs.yml
Comment (seisman) [inline ci/requirements/docs.yml:8]: Also need to update https://github.com/GenericMappingTools/pygmt/blob/main/.github/ISSUE_TEMPLATE/bump_gmt_checklist.md

Comment (weiji14) [reply ci/requirements/docs.yml:8]: Should be ok to update `bump_gmt_checklist.md` directly in #1990.

**Review by maxrjones — APPROVED** [review]
LGTM, thanks! We should update the permissions for the vercel app after merging this.

**Comment by weiji14** [comment]
Just wanted to say, thanks @maxrjones and @seisman for keeping things going! Sorry that I haven't been as active lately, it's been really intense with [FDL2022](https://frontierdevelopmentlab.org/fdl2022) (even over weekends), and I'll probably be busy until mid-August at least. Hoping to get back up to speed on PyGMT contributions from September onwards after I catch up with a backlog of stuff!

> LGTM, thanks! We should update the permissions for the vercel app after merging this.

I took at a look at the Vercel project dashboard for pygmt, and it should be disconnected already as far as I can tell (maybe someone already disconnected it in the GitHub org settings)? I didn't delete the PyGMT project on Vercel, so the old links vercel links in PRs should still work as long as they're still willing to host it. Will keep an eye out in case comments like https://github.com/GenericMappingTools/pygmt/commit/d3b5fdc3f4ea498973241ff0b6d07cf74b600b1d#commitcomment-79850583 still pop up.

**Comment by seisman** [comment]
> I took at a look at the Vercel project dashboard for pygmt, and it should be disconnected already as far as I can tell (maybe someone already disconnected it in the GitHub org settings)? 

Yes, I did it.

> Will keep an eye out in case comments like [d3b5fdc#commitcomment-79850583](https://github.com/GenericMappingTools/pygmt/commit/d3b5fdc3f4ea498973241ff0b6d07cf74b600b1d#commitcomment-79850583) still pop up.

I think the comment was made before I disconnected the Vercel service, so we should be good to go.

**Pull request merged**
